File: frontend/src/api/observability.ts
/**
 * API functions for job run tracking and data quality monitoring.
 * Provides observability into system health and job execution history.
 */

import { supabase, isSupabaseConfigured } from "../lib/supabase";

export interface JobRun {
    id: number;
    job_name: string;
    run_date: string;
    started_at: string;
    completed_at: string;
    status: "success" | "partial_failure" | "failure";
    records_processed: number;
    records_failed: number;
    duration_seconds: number;
    error_message: string | null;
    metadata: Record<string, unknown>;
    created_at: string;
}

export interface DataQualityIssue {
    id: number;
    check_date: string;
    check_type: string;
    severity: "info" | "warning" | "error";
    affected_count: number;
    affected_symbols: string[];
    description: string;
    details: Record<string, unknown>;
    resolved_at: string | null;
    created_at: string;
}

export interface JobRunStats {
    job_name: string;
    total_runs: number;
    successful_runs: number;
    failed_runs: number;
    success_rate: number;
    avg_duration_seconds: number;
    avg_records_processed: number;
    last_run_at: string;
    last_run_status: "success" | "partial_failure" | "failure";
}

/**
 * Fetch recent job runs.
 * @param days Number of days to look back (default: 30)
 * @param limit Maximum number of records (default: 100)
 */
export async function getJobRuns(
    days: number = 30,
    limit: number = 100
): Promise<JobRun[]> {
    if (!isSupabaseConfigured || !supabase) return [];

    const startDate = new Date();
    startDate.setDate(startDate.getDate() - days);

    const { data, error } = await supabase
        .from("job_runs")
        .select("*")
        .gte("run_date", startDate.toISOString().split("T")[0])
        .order("run_date", { ascending: false })
        .order("started_at", { ascending: false })
        .limit(limit);

    if (error) {
        console.error("Error fetching job runs:", error);
        return [];
    }

    return (data as JobRun[]) || [];
}

/**
 * Fetch the latest run for each job type.
 */
export async function getLatestJobRuns(): Promise<JobRun[]> {
    if (!isSupabaseConfigured || !supabase) return [];

    const { data, error } = await supabase
        .from("v_latest_job_runs")
        .select("*");

    if (error) {
        console.error("Error fetching latest job runs:", error);
        return [];
    }

    return (data as JobRun[]) || [];
}

/**
 * Fetch job run statistics.
 * @param days Number of days to aggregate (default: 30)
 */
export async function getJobRunStats(days: number = 30): Promise<JobRunStats[]> {
    if (!isSupabaseConfigured || !supabase) return [];

    // Use explicit type cast since Supabase types may not include custom functions
    const { data, error } = await supabase.rpc("get_job_run_stats", {
        p_days: days,
    } as unknown as undefined);

    if (error) {
        console.error("Error fetching job run stats:", error);
        return [];
    }

    return (data as JobRunStats[]) || [];
}

/**
 * Fetch unresolved data quality issues.
 * @param limit Maximum number of issues (default: 50)
 */
export async function getDataQualityIssues(
    limit: number = 50
): Promise<DataQualityIssue[]> {
    if (!isSupabaseConfigured || !supabase) return [];

    const { data, error } = await supabase
        .from("data_quality_checks")
        .select("*")
        .is("resolved_at", null)
        .order("check_date", { ascending: false })
        .limit(limit);

    if (error) {
        console.error("Error fetching data quality issues:", error);
        return [];
    }

    return (data as DataQualityIssue[]) || [];
}

/**
 * Fetch all data quality issues (including resolved).
 * @param days Number of days to look back (default: 30)
 * @param limit Maximum number of issues (default: 100)
 */
export async function getAllDataQualityIssues(
    days: number = 30,
    limit: number = 100
): Promise<DataQualityIssue[]> {
    if (!isSupabaseConfigured || !supabase) return [];

    const startDate = new Date();
    startDate.setDate(startDate.getDate() - days);

    const { data, error } = await supabase
        .from("data_quality_checks")
        .select("*")
        .gte("check_date", startDate.toISOString().split("T")[0])
        .order("check_date", { ascending: false })
        .limit(limit);

    if (error) {
        console.error("Error fetching all data quality issues:", error);
        return [];
    }

    return (data as DataQualityIssue[]) || [];
}

/**
 * Get symbols with stale data from the database view.
 */
export async function getStaleDataSymbols(): Promise<
    Array<{
        symbol: string;
        name: string;
        last_price_date: string | null;
        days_since_update: number | null;
        staleness_status: string;
    }>
> {
    if (!isSupabaseConfigured || !supabase) return [];

    const { data, error } = await supabase
        .from("v_stale_data_check")
        .select("symbol, name, last_price_date, days_since_update, staleness_status")
        .in("staleness_status", ["stale", "never"])
        .limit(100);

    if (error) {
        console.error("Error fetching stale data symbols:", error);
        return [];
    }

    return data || [];
}

/**
 * Get the overall system health summary.
 */
export interface SystemHealthSummary {
    lastSuccessfulRun: string | null;
    daysSinceUpdate: number;
    unresolvedIssueCount: number;
    recentJobFailures: number;
    healthStatus: "healthy" | "warning" | "error" | "unknown";
}

export async function getSystemHealthSummary(): Promise<SystemHealthSummary> {
    const defaultSummary: SystemHealthSummary = {
        lastSuccessfulRun: null,
        daysSinceUpdate: 999,
        unresolvedIssueCount: 0,
        recentJobFailures: 0,
        healthStatus: "unknown",
    };

    if (!isSupabaseConfigured || !supabase) return defaultSummary;

    try {
        // Get ingest status which includes last_successful_run
        // Use explicit type cast since Supabase types may not include updated function signature
        const { data: ingestStatus } = await supabase.rpc("get_ingest_status");
        const statusData = ingestStatus as unknown as Array<{
            last_successful_run: string | null;
            days_since_update: number;
            unresolved_issues: number;
        }> | null;
        
        // Get recent job failures (last 7 days)
        const sevenDaysAgo = new Date();
        sevenDaysAgo.setDate(sevenDaysAgo.getDate() - 7);
        
        const { data: recentFailures } = await supabase
            .from("job_runs")
            .select("id")
            .eq("status", "failure")
            .gte("run_date", sevenDaysAgo.toISOString().split("T")[0]);

        const lastSuccessfulRun = statusData?.[0]?.last_successful_run || null;
        const daysSinceUpdate = statusData?.[0]?.days_since_update || 999;
        const unresolvedIssueCount = statusData?.[0]?.unresolved_issues || 0;
        const recentJobFailures = recentFailures?.length || 0;

        // Determine health status
        let healthStatus: SystemHealthSummary["healthStatus"] = "healthy";
        if (daysSinceUpdate > 3 || recentJobFailures > 3 || unresolvedIssueCount > 10) {
            healthStatus = "error";
        } else if (daysSinceUpdate > 1 || recentJobFailures > 0 || unresolvedIssueCount > 0) {
            healthStatus = "warning";
        }

        return {
            lastSuccessfulRun,
            daysSinceUpdate,
            unresolvedIssueCount,
            recentJobFailures,
            healthStatus,
        };
    } catch (error) {
        console.error("Error fetching system health summary:", error);
        return defaultSummary;
    }
}

File: jobs/src/asx_jobs/observability.py
"""Job run tracking and data quality monitoring.

This module provides observability capabilities for the ASX Jobs Runner:
- Persists job execution metadata to the database
- Monitors data quality and detects issues
- Enables visibility into system health
"""

from datetime import datetime
from typing import Any

from asx_jobs.database import Database
from asx_jobs.jobs.base import JobResult
from asx_jobs.logging import get_logger

logger = get_logger(__name__)


class JobRunTracker:
    """Tracks job execution metadata in the database."""

    def __init__(self, db: Database) -> None:
        """Initialize the job run tracker.

        Args:
            db: Database client for persistence.
        """
        self.db = db

    def record_job_run(self, result: JobResult) -> int:
        """Record a completed job run to the database.

        Args:
            result: JobResult from an executed job.

        Returns:
            The ID of the created job_runs record.
        """
        status = self._determine_status(result)
        run_date = result.started_at.strftime("%Y-%m-%d")

        data = {
            "job_name": result.job_name,
            "run_date": run_date,
            "started_at": result.started_at.isoformat(),
            "completed_at": result.completed_at.isoformat(),
            "status": status,
            "records_processed": result.records_processed,
            "records_failed": result.records_failed,
            "duration_seconds": round(result.duration_seconds, 2),
            "error_message": result.error_message,
            "metadata": result.metadata or {},
        }

        try:
            response = self.db.client.table("job_runs").insert(data).execute()
            run_id = int(response.data[0]["id"])
            logger.info(
                "job_run_recorded",
                job_name=result.job_name,
                run_id=run_id,
                status=status,
            )
            return run_id
        except Exception as e:
            logger.error(
                "job_run_record_failed",
                job_name=result.job_name,
                error=str(e),
            )
            raise

    def _determine_status(self, result: JobResult) -> str:
        """Determine the job status based on execution result.

        Args:
            result: JobResult to evaluate.

        Returns:
            Status string: 'success', 'partial_failure', or 'failure'.
        """
        if not result.success:
            return "failure"
        if result.records_failed > 0:
            return "partial_failure"
        return "success"

    def get_last_successful_run(self, job_name: str) -> dict[str, Any] | None:
        """Get the most recent successful run for a job.

        Args:
            job_name: Name of the job to query.

        Returns:
            Job run record or None if no successful runs found.
        """
        try:
            result = (
                self.db.client.table("job_runs")
                .select("*")
                .eq("job_name", job_name)
                .eq("status", "success")
                .order("run_date", desc=True)
                .order("started_at", desc=True)
                .limit(1)
                .execute()
            )
            if result.data:
                return dict(result.data[0])
            return None
        except Exception as e:
            logger.error(
                "get_last_successful_run_failed",
                job_name=job_name,
                error=str(e),
            )
            return None

    def get_recent_runs(
        self, job_name: str | None = None, days: int = 7, limit: int = 100
    ) -> list[dict[str, Any]]:
        """Get recent job runs.

        Args:
            job_name: Optional filter by job name.
            days: Number of days to look back.
            limit: Maximum number of records to return.

        Returns:
            List of job run records.
        """
        try:
            start_date = datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)
            start_date = start_date.replace(
                day=start_date.day - days if start_date.day > days else 1
            )

            query = (
                self.db.client.table("job_runs")
                .select("*")
                .gte("run_date", start_date.strftime("%Y-%m-%d"))
                .order("run_date", desc=True)
                .order("started_at", desc=True)
                .limit(limit)
            )

            if job_name:
                query = query.eq("job_name", job_name)

            result = query.execute()
            return [dict(r) for r in result.data]
        except Exception as e:
            logger.error("get_recent_runs_failed", error=str(e))
            return []


class DataQualityMonitor:
    """Monitors data quality and records issues."""

    def __init__(self, db: Database) -> None:
        """Initialize the data quality monitor.

        Args:
            db: Database client for queries and persistence.
        """
        self.db = db
        self.tracker = JobRunTracker(db)

    def record_issue(
        self,
        check_type: str,
        severity: str,
        affected_count: int,
        affected_symbols: list[str],
        description: str,
        details: dict[str, Any] | None = None,
    ) -> int:
        """Record a data quality issue.

        Args:
            check_type: Type of check (e.g., 'stale_data', 'missing_snapshot').
            severity: Issue severity ('info', 'warning', 'error').
            affected_count: Number of affected records.
            affected_symbols: List of affected ticker symbols.
            description: Human-readable description of the issue.
            details: Additional context as JSON.

        Returns:
            The ID of the created record.
        """
        data = {
            "check_date": datetime.now().strftime("%Y-%m-%d"),
            "check_type": check_type,
            "severity": severity,
            "affected_count": affected_count,
            "affected_symbols": affected_symbols,
            "description": description,
            "details": details or {},
        }

        try:
            response = self.db.client.table("data_quality_checks").insert(data).execute()
            issue_id = int(response.data[0]["id"])
            logger.info(
                "data_quality_issue_recorded",
                check_type=check_type,
                severity=severity,
                affected_count=affected_count,
                issue_id=issue_id,
            )
            return issue_id
        except Exception as e:
            logger.error(
                "data_quality_issue_record_failed",
                check_type=check_type,
                error=str(e),
            )
            raise

    def check_stale_data(self, days_threshold: int = 7) -> dict[str, Any]:
        """Check for symbols without recent price data.

        Args:
            days_threshold: Number of days without data to consider stale.

        Returns:
            Dictionary with check results including count and symbols.
        """
        try:
            result = (
                self.db.client.table("v_stale_data_check")
                .select("symbol, days_since_update, staleness_status")
                .or_("staleness_status.eq.stale,staleness_status.eq.never")
                .execute()
            )

            stale_symbols = [row["symbol"] for row in result.data]
            count = len(stale_symbols)

            if count > 0:
                self.record_issue(
                    check_type="stale_data",
                    severity="warning" if count < 50 else "error",
                    affected_count=count,
                    affected_symbols=stale_symbols[:100],
                    description=f"{count} symbols without price data for {days_threshold}+ days",
                    details={"days_threshold": days_threshold},
                )

            logger.info(
                "stale_data_check_completed",
                stale_count=count,
            )

            return {
                "check_type": "stale_data",
                "count": count,
                "symbols": stale_symbols,
            }
        except Exception as e:
            logger.error("stale_data_check_failed", error=str(e))
            return {"check_type": "stale_data", "count": 0, "symbols": [], "error": str(e)}

    def check_missing_today_snapshot(self) -> dict[str, Any]:
        """Check for active symbols missing today's price snapshot.

        Returns:
            Dictionary with check results.
        """
        try:
            today = datetime.now().strftime("%Y-%m-%d")

            active_instruments = self.db.get_all_active_instruments()
            active_ids = {inst["id"] for inst in active_instruments}
            symbol_map = {inst["id"]: inst["symbol"] for inst in active_instruments}

            prices_today = (
                self.db.client.table("daily_prices")
                .select("instrument_id")
                .eq("trade_date", today)
                .execute()
            )
            ids_with_prices = {row["instrument_id"] for row in prices_today.data}

            missing_ids = active_ids - ids_with_prices
            missing_symbols = [symbol_map[id_] for id_ in missing_ids if id_ in symbol_map]

            count = len(missing_symbols)

            if count > 0 and count < len(active_ids) * 0.9:
                self.record_issue(
                    check_type="missing_snapshot",
                    severity="warning",
                    affected_count=count,
                    affected_symbols=missing_symbols[:100],
                    description=f"{count} active symbols missing today's price snapshot",
                    details={"check_date": today},
                )

            logger.info(
                "missing_snapshot_check_completed",
                missing_count=count,
                total_active=len(active_ids),
            )

            return {
                "check_type": "missing_snapshot",
                "count": count,
                "symbols": missing_symbols,
            }
        except Exception as e:
            logger.error("missing_snapshot_check_failed", error=str(e))
            return {"check_type": "missing_snapshot", "count": 0, "symbols": [], "error": str(e)}

    def check_price_quality(self, days: int = 7) -> dict[str, Any]:
        """Check for abnormal or invalid price values.

        Args:
            days: Number of days to check.

        Returns:
            Dictionary with check results.
        """
        try:
            result = (
                self.db.client.table("v_price_quality_issues")
                .select("symbol, issue_type, trade_date")
                .execute()
            )

            issues = result.data
            count = len(issues)
            affected_symbols = list(set(row["symbol"] for row in issues))

            if count > 0:
                issue_types: dict[str, int] = {}
                for issue in issues:
                    issue_type = issue["issue_type"]
                    issue_types[issue_type] = issue_types.get(issue_type, 0) + 1

                self.record_issue(
                    check_type="abnormal_prices",
                    severity="error" if count > 10 else "warning",
                    affected_count=count,
                    affected_symbols=affected_symbols[:100],
                    description=f"{count} price records with data quality issues",
                    details={"issue_breakdown": issue_types},
                )

            logger.info(
                "price_quality_check_completed",
                issues_found=count,
            )

            return {
                "check_type": "abnormal_prices",
                "count": count,
                "symbols": affected_symbols,
            }
        except Exception as e:
            logger.error("price_quality_check_failed", error=str(e))
            return {"check_type": "abnormal_prices", "count": 0, "symbols": [], "error": str(e)}

    def run_all_checks(self) -> dict[str, Any]:
        """Run all data quality checks.

        Returns:
            Dictionary with all check results.
        """
        logger.info("data_quality_checks_started")

        results: dict[str, Any] = {
            "timestamp": datetime.now().isoformat(),
            "checks": {},
        }

        results["checks"]["stale_data"] = self.check_stale_data()
        results["checks"]["missing_snapshot"] = self.check_missing_today_snapshot()
        results["checks"]["price_quality"] = self.check_price_quality()

        checks: dict[str, dict[str, Any]] = results["checks"]
        total_issues = sum(check.get("count", 0) for check in checks.values())
        results["total_issues"] = total_issues

        logger.info(
            "data_quality_checks_completed",
            total_issues=total_issues,
        )

        return results

    def get_unresolved_issues(self, limit: int = 100) -> list[dict[str, Any]]:
        """Get all unresolved data quality issues.

        Args:
            limit: Maximum number of issues to return.

        Returns:
            List of unresolved issue records.
        """
        try:
            result = (
                self.db.client.table("data_quality_checks")
                .select("*")
                .is_("resolved_at", "null")
                .order("check_date", desc=True)
                .limit(limit)
                .execute()
            )
            return [dict(r) for r in result.data]
        except Exception as e:
            logger.error("get_unresolved_issues_failed", error=str(e))
            return []

    def resolve_issue(self, issue_id: int) -> bool:
        """Mark a data quality issue as resolved.

        Args:
            issue_id: ID of the issue to resolve.

        Returns:
            True if successful, False otherwise.
        """
        try:
            self.db.client.table("data_quality_checks").update(
                {"resolved_at": datetime.now().isoformat()}
            ).eq("id", issue_id).execute()
            logger.info("data_quality_issue_resolved", issue_id=issue_id)
            return True
        except Exception as e:
            logger.error(
                "data_quality_issue_resolve_failed",
                issue_id=issue_id,
                error=str(e),
            )
            return False
